Update "Web embedding" documentation. (#10977)
flutter/website: src/content/platform-integration/web/embedding-flutter-web.md

  </body>
</html>
```

When Flutter web is bootstrapped without referencing `multiViewEnabled` or a 
`hostElement`, it uses full-screen mode.

See [Customize app initialization][] to learn more about the
`flutter_bootstrap.js` file.

[Customize app initialization]: {{site.docs}}/platform-integration/web/initialization/

### `iframe` embedding

Full-screen mode is recommended when embedding a Flutter web application in an
`iframe`. The page that embeds the `iframe` can size and position it as needed,
and Flutter will fill it completely.

```html
<iframe src="https://url-to-your-flutter/index.html"></iframe>
```

Check [`<iframe>`: The Inline Frame element][] in MDN to learn more about the
pros and cons of `iframe`s.

[`<iframe>`: The Inline Frame element]: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe

## Embedded mode

Flutter web applications can also render content into an arbitrary number of
elements (commonly `div`s) of another web application; this is called "embedded
mode" (or "multi-view").

In this mode:

* A Flutter web application can launch, but doesn't render until the first
"view" is added, with `addView`.

> * The host application may add or remove views from the embedded Flutter web

Author: IANAL, and this certainly isn't a legal doc :P

application.
* The Flutter application gets notified of when views are added or removed, so
it can adjust its widgets accordingly.

### Enable multi-view mode

Enable multi-view mode setting `multiViewEnabled: true` in the
`initializeEngine` method as shown:

```js highlightLines=8
// flutter_bootstrap.js
{% raw %}{{flutter_js}}{% endraw %}
{% raw %}{{flutter_build_config}}{% endraw %}

_flutter.loader.load({
  onEntrypointLoaded: async function onEntrypointLoaded(engineInitializer) {
    let engine = await engineInitializer.initializeEngine({
      multiViewEnabled: true, // Enables embedded mode.
    });
    let app = await engine.runApp();

>     // Make this `app` object available to your JS app.

Reviewer: How does the developer use `app` exactly? It's not very useful if it's only available in the scope of the `onEntrypointLoaded` callback. It seems like the developer needs to do something like:

```javascript
function getApp() {
  return new Promise((resolve) => { 
    _flutter.loader.load({
      onEntrypointLoaded: async function onEntrypointLoaded(engineInitializer) {
        let engine = await engineInitializer.initializeEngine({
          multiViewEnabled: true, 
        });
        let app = await engine.runApp();
        resolve(app); // Resolve the promise with the app
      }
    });
  });
}
```

Author: Isn't this the same, without needing a promise (that can't be re-resolved upon hot reload?):

```js
{{flutter_js}}
{{flutter_build_config}}

_flutter.loader.load({
    onEntrypointLoaded: async function onEntrypointLoaded(engineInitializer) {
        let engine = await engineInitializer.initializeEngine({
            multiViewEnabled: true,
        });
        let app = await engine.runApp();
        doThingsWithApp(app); // Pass the app to the JS that needs it.
    }
});

function doThingsWithApp(app) {
    // var app = await getApp(); // app comes as a parameter now
    let viewId1 = app.addView({
        hostElement: document.querySelector('div#host1')
    })
    let viewId2 = app.addView({
        hostElement: document.querySelector('div#host2')
    })
    console.log('view id = ' + viewId1)
    console.log('view id 2 = ' + viewId2)
}
```

I'm not sure how prescriptive we want to be with this bit of the code; it's highly dependent on the host web app, isn't it?

Reviewer: That's also a reasonable approach. I'm OK to keep this snippet as-is since it's up to the developer to decide how they want to access `app`.

  }
});
```

### Manage Flutter views from JS

To add or remove views, use the `app` object returned by the `runApp` method:

```js highlightLines=2-4,7
// Adding a view...
let viewId = app.addView({
  hostElement: document.querySelector('#some-element'),
});

// Removing viewId...
let viewConfig = flutterApp.removeView(viewId);
```

### Handling view changes from Dart

View additions and removals are surfaced to Flutter through the
[`didChangeMetrics` method][] of the `WidgetsBinding` class.

The complete list of views attached to your Flutter app is available
through the `WidgetsBinding.instance.platformDispatcher.views` iterable. These
views are of [type `FlutterView`][].

To render content into each `FlutterView`, your Flutter app needs to create a
[`View` widget][]. `View` widgets can be grouped together under a
[`ViewCollection` widget][].

> The following is an example from the _Multi View Playground_ that encapsulates

Reviewer 2: ```suggestion
The following is an example, from the _Multi View Playground_, encapsulates
```

the above in a `MultiViewApp` widget that can be used as the root widget for
your app. A [`WidgetBuilder` function][] will run for each `FlutterView`:

```dart highlightLines=25,39,46-49,56-61,72
// multi_view_app.dart

// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:ui' show FlutterView;
import 'package:flutter/widgets.dart';

/// Calls [viewBuilder] for every view added to the app to obtain the widget to
/// render into that view. The current view can be looked up with [View.of].

> class MultiViewApp extends StatefulWidget {

Reviewer 3: This seems SUPER complex. It feels like there should be a widget we ship here to do most of this heavy lifting.

Author: `MultiViewApp` is (one of the) widgets you're describing; this is how it's used:

* https://github.com/goderbauer/mvp/blob/main/lib/widgets_counter.dart#L9-L11

(The `viewBuilder` function can have additional complexity to decide what widgets go into each `viewId`)

  const MultiViewApp({super.key, required this.viewBuilder});

  final WidgetBuilder viewBuilder;

  @override
  State<MultiViewApp> createState() => _MultiViewAppState();
}

class _MultiViewAppState extends State<MultiViewApp> with WidgetsBindingObserver {
  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
    _updateViews();
  }

  @override
  void didUpdateWidget(MultiViewApp oldWidget) {
    super.didUpdateWidget(oldWidget);
    // Need to re-evaluate the viewBuilder callback for all views.
    _views.clear();
    _updateViews();
  }

  @override
  void didChangeMetrics() {
    _updateViews();
  }

>   Map<Object, Widget> _views = <Object, Widget>{};

Reviewer 3: ```suggestion
  final Map<Object, Widget> _views = <Object, Widget>{};
```

Author: Can't make this one `final`, it's re-assigned in the `setState()` call of `_updateViews()`.

    }
    setState(() {
      _views = newViews;
    });
  }

  Widget _createViewWidget(FlutterView view) {
    return View(
      view: view,
      child: Builder(
        builder: widget.viewBuilder,
      ),
    );
  }

  @override
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ViewCollection(views: _views.values.toList(growable: false));
  }
}
```

For more information, check out [`WidgetsBinding` mixin][] in the API docs, or
the [Multi View Playground repo][] that was used during development.

[`didChangeMetrics` method]: {{site.api}}/flutter/widgets/WidgetsBindingObserver/didChangeMetrics.html
[Multi View Playground repo]: {{site.github}}/goderbauer/mvp
[type `FlutterView`]: {{site.api}}/flutter/dart-ui/FlutterView-class.html
[`View` widget]: {{site.api}}/flutter/widgets/View-class.html
[`ViewCollection` widget]: {{site.api}}/flutter/widgets/ViewCollection-class.html
[`WidgetsBinding` mixin]: {{site.api}}/flutter/widgets/WidgetsBinding-mixin.html
[`WidgetBuilder` function]: {{site.api}}/flutter/widgets/WidgetBuilder.html

### Identifying views

> Each `FlutterView` has an identifier that is assigned by Flutter when it is
> attached. This `viewId` can be used to uniquely identify each view, retrieve

Reviewer 2: ```suggestion
Each `FlutterView` has an identifier assigned by Flutter when
attached. This `viewId` can be used to uniquely identify each view, retrieve
```

its initial configuration, or decide what to render in it.

> The `viewId` of the `FlutterView` that is being rendered can be retrieved from

Reviewer 2: ```suggestion
The `viewId` of the rendered `FlutterView` can be retrieved from
```

method, as shown:

```js highlightLines=4-7
// Adding a view with initial data...
let viewId = app.addView({
  hostElement: someElement,
  initialData: {
    greeting: 'Hello, world!',
    randomValue: Math.floor(Math.random() * 100),
  }
});
```

In Dart, the `initialData` is available as a `JSAny` object, accessible through
the top-level `views` property in the `dart:ui_web` library. The data is
accessed through the `viewId` of the current view, like so: 

```dart
final initialData = ui_web.views.getInitialData(viewId) as YourJsInteropType;
```

To learn how to define the `YourJsInteropType` class to map the `initialData`
object passed from JS so it's type-safe in your Dart program, check out:
[JS Interoperability][] on dart.dev.

[JS Interoperability]: {{site.dart-site}}/interop/js-interop

### View constraints

By default, an embedded Flutter web view considers the size of its `hostElement`
as an immutable property, and will tightly constrain its layout to the available
space.

On the web, it's common for the intrinsic size of an element to affect the
layout of the page (like an `img` tag or `p` tags that can reflow content around
them).

> When adding a view to Flutter web, you may configure it with constraints that

Author: Should this be `can` as well?

inform Flutter of how the view needs to be laid out:

```js highlightLines=4-8
// Adding a view with initial data...
let viewId = app.addView({
  hostElement: someElement,
  viewConstraints: {
    maxWidth: 320,
    minHeight: 0,
    maxHeight: Infinity,
  }
});
```

The view constraints passed from JS need to be compatible with the CSS styling
of the `hostElement` where Flutter is being embedded. For example, Flutter
won’t try to "fix" contradictory constants like passing  `max-height: 100px`
in CSS, but `maxHeight: Infinity` to Flutter.

Learn more about the [`ViewConstraints` class][], and
[Understanding constraints][].

[`ViewConstraints` class]: {{site.api}}/flutter/dart-ui/ViewConstraints-class.html
[Understanding constraints]: {{site.docs}}/ui/layout/constraints

## Custom element (`hostElement`)

> _Between Flutter 3.10 and 3.24_<br />

Reviewer 2: Will this be true in 3.25? If so, we should say "In Flutter 3.10 and beyond..."

Author: This will be still true, but starting in 3.24 people should prefer the other embedded method

Reviewer 2: I see! OK, I'm going to make another suggested change.

You can embed a single-view Flutter web app into any HTML element of your web
page.

> To tell Flutter web in which element to render, pass an object with a `config`

Author: _"which element to render into"_? What flutter renders and *where* it renders it are two different things. How to tell those two apart?

> To tell Flutter web in which element to render, pass an object with a `config`

Author: Yep, this sounds right to me!

field to the `_flutter.loader.load` function that specifies a `HTMLElement` as
the `hostElement`.

```js highlightLines=3
_flutter.loader.load({
  config: {
    hostElement: document.getElementById('flutter_host'),
  }
});
```

To learn more about other configuration options,
check out [Customizing web app initialization][].

:::version-note
This method of specifying the `hostElement` is superseded by the

> **Embedded mode** described above, please consider migrating to it.

Reviewer 2: ```suggestion
**Embedded mode** described above, **please consider migrating to it**.
```

> :::version-note
> This method of specifying the `hostElement` is superseded by the
> **Embedded mode** described above, **please consider migrating to it**.
> To learn how to configure the `hostElement` in earlier Flutter versions,
> reference [Legacy web app initialization][].
> :::

Reviewer 2: ```suggestion
#### Flutter 3.24 and beyond

As of the 3.24 release, specifying a custom element by
using **Embedded mode** (described above),
supersedes specifying the `hostElement` directly.
Please migrate to the Embedded mode approach.

To learn how to configure the `hostElement` in Flutter versions before 3.24,
check out [Legacy web app initialization][].
```

Author: > why not document embedded mode _first_ and then mention hostname second.

Yep, that's exactly what we're doing; check it out here:

https://flutter-docs-prod--pr10977-web-embedding-docs-o3h28ltv.web.app/platform-integration/web/embedding-flutter-web#embedded-mode

Reviewer 2: D'OH!
